### PR TITLE
🐛 use the kubernetes decoder in the patch engine

### LINKED ISSUE #6667: `BeforeClusterUpgrade` and `AfterControlPlaneUpgrade` are getting called multiple times per reconcile
**Details:**

If a cluster upgrade is blocked by the `BeforeClusterUpgrade` (or subsequently by the `AfterControlPlaneUpgrade` hook) the reconciler is triggered more than once in quick successions.  

Example: As long as the `BeforeClusterUpgrade` hook returns a hook response with `RetryAfterSeconds` set to `10` the cluster object should be requeued and processed once every ~10 seconds. But the cluster is being processed a couple of times in quick succession before is it reprocessed after 10 seconds.

**What did you expect to happen:**

The cluster object should only be processed once with in the 10 seconds.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]
 

/area runtime-sdk

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

This PR switches to using the kubernetes decoder instead of the standard yaml library to Unmarshal.

Without this change we get additional updates with SSA. 

The problem is that the Go unmarshaller decodes `1` into a float64, while the Kubernetes decoder decodes it into an int64. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6667

### CODE REVIEW

File: internal/controllers/topology/cluster/patches/template.go [hunk 1]
```diff
@@ -135,15 +136,11 @@ func getRequestItem(req *runtimehooksv1.GeneratePatchesRequest, holderKind, hold
 // bytesToUnstructured provides a utility method that converts a (JSON) byte array into an Unstructured object.
 func bytesToUnstructured(b []byte) (*unstructured.Unstructured, error) {
 	// Unmarshal the JSON.
-	var m map[string]interface{}
-	if err := json.Unmarshal(b, &m); err != nil {
+	u := &unstructured.Unstructured{}
+	if _, _, err := serializer.NewCodecFactory(nil).UniversalDeserializer().Decode(b, nil, u); err != nil {
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/patches/template.go:140]: I'm not sure but I think creating a new codec factory and a new deserializer for every bytesToUnstructured is too much.

Let's use a package global variable for the deserializer that we initialize with `init`. I think this should be okay, adding it to the engine and passing it through might be too much effort.

Comment (sbueringer) [reply internal/controllers/topology/cluster/patches/template.go:140]: \* don't need init works inline in the var declaration

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**